fe2-android/app/src/main/java/com/github/dedis/popstellar/utility/security/KeyManager.java

```diff
@@ -22,46 +26,136 @@
 import java.security.GeneralSecurityException;
 import java.util.Arrays;
 import java.util.Base64;
+import java.util.Comparator;
+import java.util.Optional;
 
 import javax.inject.Inject;
 import javax.inject.Singleton;
 
+/** Service manging keys and providing easy access to the main device key and PoP Tokens */
 @Singleton
 public class KeyManager {
 
   private static final String TAG = KeyManager.class.getSimpleName();
 
   private final AndroidKeysetManager keysetManager;
   private final Wallet wallet;
+  private KeyPair keyPair;
 
   @Inject
   public KeyManager(AndroidKeysetManager keysetManager, Wallet wallet) {
     this.keysetManager = keysetManager;
     this.wallet = wallet;
 
     try {
+      regenerateMainKey();
       Log.d(TAG, "Public Key = " + getMainPublicKey().getEncoded());
     } catch (IOException | GeneralSecurityException e) {
-      Log.e(TAG, "Failed to retrieve public key", e);
-      throw new IllegalStateException("Failed to retrieve public key", e);
+      Log.e(TAG, "Failed to retrieve device's key", e);
+      throw new IllegalStateException("Failed to retrieve device's key", e);
     }
   }
 
-  public PublicKey getMainPublicKey() throws IOException, GeneralSecurityException {
-    return getPublicKey(keysetManager.getKeysetHandle());
+  /**
+   * This will regenerate the cached device KeyPair.
+   *
+   * <p>Use this only if you know what you are doing
+   *
+   * @throws IOException when the key cannot be retrieved due to IO errors
+   * @throws GeneralSecurityException when the retrieved key is not valid
+   */
+  public void regenerateMainKey() throws GeneralSecurityException, IOException {
+    keyPair = getKeyPair(keysetManager.getKeysetHandle());
   }
 
-  public KeyPair getMainKeyPair() throws IOException, GeneralSecurityException {
-    PrivateKey privateKey = new ProtectedPrivateKey(keysetManager.getKeysetHandle());
-    PublicKey publicKey = getMainPublicKey();
+  /** @return the device public key */
+  public PublicKey getMainPublicKey() {
+    return keyPair.getPublicKey();
+  }
 
-    return new KeyPair(privateKey, publicKey);
+  /** @return the deice keypair */
+  public KeyPair getMainKeyPair() {
+    return keyPair;
   }
 
-  public PoPToken getPoPToken(String laoID, String rollCallID) throws GeneralSecurityException {
+  /**
+   * Generate the PoPToken for the given lao and roll call
+   *
+   * <p>Exceptions are thrown if the process fails
+   *
+   * @param laoID id of the lao
+   * @param rollCallID id of the lao's roll call
+   * @return the generated PoPToken
+   * @throws GeneralSecurityException if something goes wrong during the generation
+   */
+  public PoPToken requirePoPToken(String laoID, String rollCallID) throws GeneralSecurityException {
     return wallet.findKeyPair(laoID, rollCallID);
```

Comment:
This looks like a more accurate name would be 

```suggestion
    return wallet.generatePoPToken(laoID, rollCallID);
```

though that would entail changes on the wallet side.